### PR TITLE
Automate the generation of release change log draft

### PR DESCRIPTION
**Pull request by daxian-dbw**
## PR Summary
Automate the generation of release change log draft.

For a commit that is breaking change, `Get-ChangeLog` assumes that a `[breaking change]` tag is put in the commit description body so that it can incorporate this information in the generated changelog entry.

So to @PowerShell/powershell-maintainers, please put the `[breaking change]` tag in the description body when you squash merge a PR that is a breaking change.

## PR Checklist

Note: Please mark anything not applicable to this PR `NA`.

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [x] Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [NA] User facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [NA] Issue filed - Issue link:
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [NA] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
    - [NA] [Add `[feature]` if the change is significant or affectes feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.

### CODE REVIEW

File: tools/releaseTools.psm1 [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        ## We can get all new commits using the revision range "$tag_hash..HEAD", meaning the commits that are
+        ## reachable from 'HEAD' but not reachable from the last release tag.
+        $new_commits = git --no-pager log --no-merges "$tag_hash..HEAD" --format=$format | New-CommitNode
+    }
+
+    foreach ($commit in $new_commits) {
+        if ($commit.AuthorEmail.EndsWith("@microsoft.com")) {
+            $commit.ChangeLogMessage = "- {0}" -f $commit.Subject
+        } else {
+            $uri = "https://api.github.com/repos/PowerShell/PowerShell/commits/$($commit.Hash)"
```
Comment (TravisEz13) [inline tools/releaseTools.psm1]: This belongs better in this module: https://github.com/PowerShell/PowerShellForGitHub

File: tools/releaseTools.psm1 [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        } else {
+            throw "PR number is missing from the commit subject. (commit: $hash)"
+        }
+    }
+}
+
+##############################
+#.SYNOPSIS
+#In the release workflow, the release branch will be merged back to master after the release is done,
+#and a merge commit will be greated as the child of the release tag commit.
```
Comment (iSazonov) [inline tools/releaseTools.psm1]: Typo greated -> treated .

File: tools/releaseTools.psm1 (hunk 1 above)
Comment (daxian-dbw) [reply tools/releaseTools.psm1]: That module is more about general GitHub operations, while this one is specific to powershell. It's based on how we operate the repository, for example
- we use squash merge almost all the time, so we can ignore merge commits when generating changelog.
- our release process, including how we do cherry-picks and merge release branch to master.
- the format of changelog entries, as well as the assumption that `[breaking change]` tag will be put in commit description body by maintainers.

**Comment by iSazonov** [comment]
We have three places where we mark a breaking change - PR description, PR label and now commit description. Can we reduce this amount? Can we use the Breaking change PR label in the script?

File: tools/releaseTools.psm1 (hunk 1 above)
Comment (TravisEz13) [reply tools/releaseTools.psm1]: You could also optimize this by getting the author email address locally and only going to github if you don't know the login for that email.

**Review by TravisEz13 — APPROVED** [review]
Please file issues, respond to or add todo comments for comments

File: tools/releaseTools.psm1 (hunk 1 above)
Comment (daxian-dbw) [reply tools/releaseTools.psm1]: Good idea. We don't have to query GitHub for every commit from external contributors, as they may be from the same one. Here is an example of output:
```
- Replace Remaining HttpBin.org Tests with WebListener (#5665) (Thanks @markekraus!)
- Combine Web Cmdlet Partial Class Files (#5612) (Thanks @markekraus!)
- Enable conversions from PSMethod to Delegate (#5287) (Thanks @powercode!)
- Fix PSVersion in PSSessionConfiguration tests (#5554) (Thanks @iSazonov!)
- Update target tag for master CI to 6.1.0-preview.1 (#5513)
- Remove AddTypeCommandBase class (#5407) (Thanks @iSazonov!)
```
We can save 2 queries in this example.

File: tools/releaseTools.psm1 (hunk 2 above)
Comment (daxian-dbw) [reply tools/releaseTools.psm1]: Fixed. Thanks!

File: tools/releaseTools.psm1 (hunk 1 above)
Comment (daxian-dbw) [reply tools/releaseTools.psm1]: Added `$community_login_map` to cache the login names we queried.

**Review by TravisEz13 — APPROVED** [review]

**Comment by iSazonov** [comment]
@daxian-dbw Thanks for clarify. I agree that keeping the breaking change label in commit description is useful. 
Continuing to think about simplifying this process  - can we assign a template to commit description in GitHub and/or Git?

**Review by iSazonov — APPROVED** [review]

**Comment by daxian-dbw** [comment]
@iSazonov Thank you for the research! Let me play with the hook.
I'm also planning to update the maintainer README.md about the commit message requirement.

**Comment by iSazonov** [comment]
@daxian-dbw If you plan to play with Git hooks please take attention on #3504 too.

**Comment by daxian-dbw** [comment]
@iSazonov Will do, thanks!

**Comment by TravisEz13** [comment]
We should investigate what git hooks can do for us, but I cannot find any docs that GitHub supports these.  I doubt that these would be supported as it would allow fairly arbitrary code to run on GitHub's servers.

**Pull request merged**